### ISSUE TITLE
Repeated string equate causes an infinite loop when referenced

### ISSUE THREAD

**Issue by yenatch** [rednex/rgbds#64]
``` asm
string equs "anything"
string equs "anything"
string
```

rgbasm will accept the following without complaint if `string` is not referenced:

``` asm
string equs "anything"
string equs "anything"
```

If the equates don't match, rgbasm will raise `Buffer safety margin exceeded`:

``` asm
string equs "anything"
string equs "anything else"
```

**Comment by yenatch** [comment]
Oops, this is not really a bug (except for the circular reference causing an infinite).

**Comment by AntonioND** [comment]
It took me a while to understand the problem.
```
string equs "anything"
string equs "anything"
```
That is converted to:
```
string equs "anything"
anything equs "anything"
```
If `string` is used, it will be replaced by `anything`, which will be replaced by `anything` again, and so on. That's why the bug only happens when you actually use `string`. The third case is the same thing, I guess that the line `yyunputstr(s = sym_GetStringValue(dest))` is executed in an infinite loop causing the buffer to be filled (and that's why that error message is showing).

**Comment by AntonioND** [comment]
Maybe a warning? Something like "Self reference in an EQUS, it may cause an infinite loop." the is only shown once per execution of rbgasm.

**Comment by Ben10do** [comment]
Hmm... I'm struggling to think of any reasons why someone would deliberately use `EQUS` recursively. As a result, I'd say we should show an error and stop straightaway, to avoid a (most likely undesired) infinite loop.

As for detecting it, we could keep track of all the `EQUS`es we've used in the current evaluation (e.g. with a stack), and stop if we encounter an `EQUS` we've already seen. This could be a faff, but it'd solve the issue (unless if I've overlooked something).

**Comment by AntonioND** [comment]
Well, an EQUS can be assigned a multiline string. You can do something like:
```
STORE_ARGS EQUS "IF _NARG > 1\n DB \\1\n SHIFT\n STORE_ARGS\nENDC"
```
And you could use it to store an arbitrary number of macro arguments. You can't do that with a macro inside a macro.

**Comment by AntonioND** [comment]
On the other hand, it is true that some of the most hacky constructs in C are done using `#define`, so prohibiting recursive EQUS could make sense. I'm not sure about this.

**Comment by AntonioND** [comment]
By the way, the same problem exists with macros:
```
HELLO : MACRO                                                                                                         
    HELLO
ENDM

    HELLO
```

**Comment by AntonioND** [comment]
I think that the only thing that can realistically be done is to warn about this problem.

I don't think there's a good way of solving this. Both EQUS and macros can have this problem, and it can be hidden very easily if the coder wants to (by having a chain of 2 or 3 definitions, for example).

It cannot be detected by the parser either because when it finds a symbol that has to be replaced by an EQUS it simply sends the string to the lexer for it to handle. It cannot even be detected by preventing the replacement of the same EQUS in the same line twice, you could be trying to use it in two different places and that's not a problem.

Also, it can only be detected when it is actually used: An EQUS may change if you redefine it.

Macros are even worse. If you hide the recursive usage of the macro inside an IF, how can you even know if the problem is going to happen or not? You could even do crazy things like getting the assembler into a loop that only ends when a specific time is reached! xD

**Comment by AntonioND** [comment]
I've pushed this: https://github.com/rednex/rgbds/commit/4228e3e8900af0d42e5e2384af90deeef8882008

I really don't know if it makes sense to spend time trying to detect this considering that any possible test is going to be incomplete and, in the end, everything comes down to the programmer doing things right..

**Comment by NieDzejkob** [comment]
I would say to not expand macros nor equs when used as a macro or equ name

**Comment by AntonioND** [comment]
I don't think I understand what you mean, could you show what you mean with an example?

**Comment by NieDzejkob** [comment]
When defining an EQU, the part on the left of it should not be fed into the macro expanding part. That way, `string equs "anything"
string equs "anything"` will stay that way and just redefine string. Same goes for macros - you probably won't need any macros or similar in the name part.

**Comment by AntonioND** [comment]
That doesn't help solve the problem. You can still do this:
```
AAA EQUS "BBB"
BBB EQUS "AAA"

AAA ; Infinite loop
```

**Comment by ISSOtm** [comment]
A possible solution would be to enforce a recursion depth limit. This limit could be passed as a command-line argument, too, so a recursion-heavy program could be still made working.

**Comment by ISSOtm** [comment]
Fixed in e0e8170fe6bbbb786e407a9a95f5848ab10a1f26
